File: library/src/main/java/me/proxer/library/api/NotificationInfoAdapter.java
package me.proxer.library.api;

import com.squareup.moshi.FromJson;
import com.squareup.moshi.JsonDataException;
import me.proxer.library.entitiy.notifications.NotificationInfo;

/**
 * @author Ruben Gees
 */
class NotificationInfoAdapter {

    private static final int FIELD_AMOUNT = 6;

    private static final int MESSAGE_FIELD_LOCATION = 2;
    private static final int FRIEND_REQUEST_FIELD_LOCATION = 3;
    private static final int NEWS_FIELD_LOCATION = 4;
    private static final int NOTIFICATIONS_FIELD_LOCATION = 5;

    @FromJson
    NotificationInfo fromJson(final int[] json) {
        if (json.length != FIELD_AMOUNT) {
            throw new JsonDataException("Invalid json array size: " + json.length);
        }

        return new NotificationInfo(json[MESSAGE_FIELD_LOCATION], json[FRIEND_REQUEST_FIELD_LOCATION],
                json[NEWS_FIELD_LOCATION], json[NOTIFICATIONS_FIELD_LOCATION]);
    }
}

File: library/src/main/kotlin/me/proxer/library/internal/adapter/NotificationInfoAdapter.kt
package me.proxer.library.internal.adapter

import com.squareup.moshi.FromJson
import com.squareup.moshi.JsonDataException
import me.proxer.library.entity.notifications.NotificationInfo

/**
 * @author Ruben Gees
 */
internal class NotificationInfoAdapter {

    private companion object {
        private const val FIELD_AMOUNT = 6

        private const val MESSAGE_FIELD_LOCATION = 2
        private const val FRIEND_REQUEST_FIELD_LOCATION = 3
        private const val NEWS_FIELD_LOCATION = 4
        private const val NOTIFICATIONS_FIELD_LOCATION = 5
    }

    @FromJson
    fun fromJson(json: IntArray): NotificationInfo {
        if (json.size != FIELD_AMOUNT) {
            throw JsonDataException("Invalid json array size: " + json.size)
        }

        return NotificationInfo(
            json[MESSAGE_FIELD_LOCATION], json[FRIEND_REQUEST_FIELD_LOCATION],
            json[NEWS_FIELD_LOCATION], json[NOTIFICATIONS_FIELD_LOCATION]
        )
    }
}
